File: kotlin-sdk/src/main/kotlin/com/cleanster/api/PropertiesApi.kt
package com.cleanster.api

import com.cleanster.CleansterClient
import com.cleanster.model.*

/** API methods for managing cleaning properties (locations). */
class PropertiesApi internal constructor(private val client: CleansterClient) {

    /** List all properties. Optionally filter by service type. */
    suspend fun listProperties(serviceId: Int? = null): ApiResponse<List<Any>> = client.request(
        method      = "GET",
        path        = "/v1/properties",
        queryParams = serviceId?.let { mapOf("serviceId" to "$it") } ?: emptyMap(),
    )

    /** Create a new property. */
    suspend fun addProperty(request: CreatePropertyRequest): ApiResponse<Property> = client.request(
        method = "POST",
        path   = "/v1/properties",
        body   = request,
    )

    /** Get a single property by ID. */
    suspend fun getProperty(propertyId: Int): ApiResponse<Property> = client.request(
        method = "GET",
        path   = "/v1/properties/$propertyId",
    )

    /** Update an existing property. */
    suspend fun updateProperty(propertyId: Int, request: CreatePropertyRequest): ApiResponse<Property> = client.request(
        method = "PUT",
        path   = "/v1/properties/$propertyId",
        body   = request,
    )

    /** Permanently delete a property. */
    suspend fun deleteProperty(propertyId: Int): ApiResponse<Map<String, Any>> = client.request(
        method = "DELETE",
        path   = "/v1/properties/$propertyId",
    )

    /** Update freeform additional information fields on a property. */
    suspend fun updateAdditionalInformation(
        propertyId: Int,
        fields:     Map<String, String>,
    ): ApiResponse<Map<String, Any>> = client.request(
        method = "PUT",
        path   = "/v1/properties/$propertyId/additional-information",
        body   = fields,
    )

    /** Enable or disable a property. Disabled properties cannot receive new bookings. */
    suspend fun enableOrDisableProperty(propertyId: Int, enabled: Boolean): ApiResponse<Map<String, Any>> = client.request(
        method = "POST",
        path   = "/v1/properties/$propertyId/enable-disable",
        body   = EnableDisablePropertyRequest(enabled = enabled),
    )

    /** List the cleaners associated with a property. */
    suspend fun getPropertyCleaners(propertyId: Int): ApiResponse<List<Any>> = client.request(
        method = "GET",
        path   = "/v1/properties/$propertyId/cleaners",
    )

    /** Add a cleaner to a property's preferred pool. */
    suspend fun addCleanerToProperty(propertyId: Int, cleanerId: Int): ApiResponse<Map<String, Any>> = client.request(
        method = "POST",
        path   = "/v1/properties/$propertyId/cleaners",
        body   = AddPropertyCleanerRequest(cleanerId = cleanerId),
    )

    /** Remove a cleaner from a property's preferred pool. */
    suspend fun removeCleanerFromProperty(propertyId: Int, cleanerId: Int): ApiResponse<Map<String, Any>> = client.request(
        method = "DELETE",
        path   = "/v1/properties/$propertyId/cleaners/$cleanerId",
    )

    /** Add one or more iCal feed URLs for calendar sync (e.g. Airbnb, VRBO). Each URL must be a live, publicly fetchable .ics feed. */
    suspend fun setICalLink(propertyId: Int, calendarLinks: List<String>): ApiResponse<Map<String, Any>> = client.request(
        method = "PUT",
        path   = "/v1/properties/$propertyId/ical",
        body   = SetICalLinkRequest(calendarLinks = calendarLinks),
    )

    /** Retrieve the calendar links currently attached to a property. */
    suspend fun getICalLink(propertyId: Int): ApiResponse<Map<String, Any>> = client.request(
        method = "GET",
        path   = "/v1/properties/$propertyId/ical",
    )

    /** Remove one or more calendar links from a property, by numeric link ID (from getICalLink) - not by URL. */
    suspend fun deleteICalLink(propertyId: Int, ids: List<Int>): ApiResponse<Map<String, Any>> = client.request(
        method = "DELETE",
        path   = "/v1/properties/$propertyId/ical",
        body   = DeleteICalLinkRequest(ids = ids),
    )

    /**
     * Assign a checklist as the default for all future bookings on a property.
     *
     * @param updateUpcomingBookings If `true`, also updates already-scheduled bookings.
     */
    suspend fun setDefaultChecklist(
        propertyId:             Int,
        checklistId:            Int,
        updateUpcomingBookings: Boolean = false,
    ): ApiResponse<Map<String, Any>> = client.request(
        method      = "PUT",
        path        = "/v1/properties/$propertyId/checklist/$checklistId",
        queryParams = mapOf("updateUpcomingBookings" to "$updateUpcomingBookings"),
    )
}

File: swift-sdk/Sources/Cleanster/Api/PropertiesApi.swift
import Foundation

/// API methods for managing cleaning properties (locations).
public final class PropertiesApi {
    private let client: CleansterClient
    init(client: CleansterClient) { self.client = client }

    /// List all properties. Optionally filter by service type.
    public func listProperties(serviceId: Int? = nil) async throws -> ApiResponse<AnyCodable> {
        var query: [URLQueryItem]? = nil
        if let sid = serviceId {
            query = [URLQueryItem(name: "serviceId", value: "\(sid)")]
        }
        return try await client.requestRaw(method: "GET", path: "/v1/properties", queryItems: query)
    }

    /// Create a new property.
    public func addProperty(_ request: CreatePropertyRequest) async throws -> ApiResponse<Property> {
        return try await client.request(method: "POST", path: "/v1/properties", body: request)
    }

    /// Get a single property by ID.
    public func getProperty(_ propertyId: Int) async throws -> ApiResponse<Property> {
        return try await client.request(method: "GET", path: "/v1/properties/\(propertyId)")
    }

    /// Update an existing property.
    public func updateProperty(_ propertyId: Int, request: CreatePropertyRequest) async throws -> ApiResponse<Property> {
        return try await client.request(method: "PUT", path: "/v1/properties/\(propertyId)", body: request)
    }

    /// Permanently delete a property.
    public func deleteProperty(_ propertyId: Int) async throws -> ApiResponse<AnyCodable> {
        return try await client.requestRaw(method: "DELETE", path: "/v1/properties/\(propertyId)")
    }

    /// Update freeform additional information fields on a property.
    public func updateAdditionalInformation(
        _ propertyId: Int,
        fields: [String: String]
    ) async throws -> ApiResponse<AnyCodable> {
        let body = UpdateAdditionalInfoRequest(fields: fields)
        return try await client.requestRaw(
            method: "PUT",
            path: "/v1/properties/\(propertyId)/additional-information",
            body: body
        )
    }

    /// Enable or disable a property. Disabled properties cannot receive new bookings.
    public func enableOrDisableProperty(_ propertyId: Int, enabled: Bool) async throws -> ApiResponse<AnyCodable> {
        let body = EnableDisablePropertyRequest(enabled: enabled)
        return try await client.requestRaw(
            method: "POST",
            path: "/v1/properties/\(propertyId)/enable-disable",
            body: body
        )
    }

    /// List the cleaners associated with a property.
    public func getPropertyCleaners(_ propertyId: Int) async throws -> ApiResponse<AnyCodable> {
        return try await client.requestRaw(method: "GET", path: "/v1/properties/\(propertyId)/cleaners")
    }

    /// Add a cleaner to a property's preferred pool.
    public func addCleanerToProperty(_ propertyId: Int, cleanerId: Int) async throws -> ApiResponse<AnyCodable> {
        let body = AddPropertyCleanerRequest(cleanerId: cleanerId)
        return try await client.requestRaw(
            method: "POST",
            path: "/v1/properties/\(propertyId)/cleaners",
            body: body
        )
    }

    /// Remove a cleaner from a property's preferred pool.
    public func removeCleanerFromProperty(_ propertyId: Int, cleanerId: Int) async throws -> ApiResponse<AnyCodable> {
        return try await client.requestRaw(
            method: "DELETE",
            path: "/v1/properties/\(propertyId)/cleaners/\(cleanerId)"
        )
    }

    /// Set an iCal feed URL for calendar sync (e.g. Airbnb, VRBO).
    public func setICalLink(_ propertyId: Int, icalLink: String) async throws -> ApiResponse<AnyCodable> {
        let body = SetICalLinkRequest(icalLink: icalLink)
        return try await client.requestRaw(
            method: "PUT",
            path: "/v1/properties/\(propertyId)/ical",
            body: body
        )
    }

    /// Retrieve the current iCal feed URL for a property.
    public func getICalLink(_ propertyId: Int) async throws -> ApiResponse<AnyCodable> {
        return try await client.requestRaw(method: "GET", path: "/v1/properties/\(propertyId)/ical")
    }

    /// Remove the iCal feed link from a property.
    public func deleteICalLink(_ propertyId: Int, icalLink: String) async throws -> ApiResponse<AnyCodable> {
        let body = DeleteICalLinkRequest(icalLink: icalLink)
        return try await client.requestRaw(
            method: "DELETE",
            path: "/v1/properties/\(propertyId)/ical",
            body: body
        )
    }

    /// Assign a checklist as the default for all future bookings on a property.
    ///
    /// - Parameter updateUpcomingBookings: If `true`, also updates already-scheduled bookings.
    public func setDefaultChecklist(
        _ propertyId: Int,
        checklistId: Int,
        updateUpcomingBookings: Bool = false
    ) async throws -> ApiResponse<AnyCodable> {
        let query = [URLQueryItem(name: "updateUpcomingBookings", value: "\(updateUpcomingBookings)")]
        return try await client.requestRaw(
            method: "PUT",
            path: "/v1/properties/\(propertyId)/checklist/\(checklistId)",
            queryItems: query
        )
    }
}
